File: trabalho_3/cliente_java/src/models/Clube.java
package models;

/**
 * Classe que representa um clube de futebol
 */
public class Clube {
    private String nome;
    private String sigla;
    private int vitorias;
    private int empates;
    private int derrotas;
    private int golsPro;
    private int golsContra;

    // Construtores
    public Clube() {
        this("", "", 0, 0, 0, 0, 0);
    }

    public Clube(String nome, String sigla) {
        this(nome, sigla, 0, 0, 0, 0, 0);
    }

    public Clube(String nome, String sigla, int vitorias, int empates, int derrotas, int golsPro, int golsContra) {
        this.nome = nome;
        this.sigla = sigla;
        this.vitorias = vitorias;
        this.empates = empates;
        this.derrotas = derrotas;
        this.golsPro = golsPro;
        this.golsContra = golsContra;
    }

    // Getters e Setters
    public String getNome() { return nome; }
    public void setNome(String nome) { this.nome = nome; }

    public String getSigla() { return sigla; }
    public void setSigla(String sigla) { this.sigla = sigla; }

    public int getVitorias() { return vitorias; }
    public void setVitorias(int vitorias) { this.vitorias = vitorias; }

    public int getEmpates() { return empates; }
    public void setEmpates(int empates) { this.empates = empates; }

    public int getDerrotas() { return derrotas; }
    public void setDerrotas(int derrotas) { this.derrotas = derrotas; }

    public int getGolsPro() { return golsPro; }
    public void setGolsPro(int golsPro) { this.golsPro = golsPro; }

    public int getGolsContra() { return golsContra; }
    public void setGolsContra(int golsContra) { this.golsContra = golsContra; }

    // Métodos de cálculo
    public int getTotalJogos() {
        return vitorias + empates + derrotas;
    }

    public int getPontos() {
        return (vitorias * 3) + empates;
    }

    public int getSaldoGols() {
        return golsPro - golsContra;
    }

    // Métodos de conversão JSON
    public String toJson() {
        return String.format(
            "{\"nome\":\"%s\", \"sigla\":\"%s\", \"vitorias\":%d, \"empates\":%d, \"derrotas\":%d, \"gols_pro\":%d, \"gols_contra\":%d}",
            nome, sigla, vitorias, empates, derrotas, golsPro, golsContra
        );
    }

    /**
     * Cria um objeto Clube a partir de uma string JSON simples
     * Nota: Esta é uma implementação simplificada para evitar dependências externas
     */
    public static Clube fromJson(String json) {
        Clube clube = new Clube();
        
        // Parse simples do JSON (sem biblioteca externa)
        json = json.trim();
        if (json.startsWith("{") && json.endsWith("}")) {
            json = json.substring(1, json.length() - 1);
            String[] pairs = json.split(",");
            
            for (String pair : pairs) {
                String[] keyValue = pair.split(":", 2);
                if (keyValue.length == 2) {
                    String key = keyValue[0].trim().replaceAll("\"", "");
                    String value = keyValue[1].trim().replaceAll("\"", "");
                    
                    switch (key) {
                        case "nome":
                            clube.setNome(value);
                            break;
                        case "sigla":
                            clube.setSigla(value);
                            break;
                        case "vitorias":
                            clube.setVitorias(Integer.parseInt(value));
                            break;
                        case "empates":
                            clube.setEmpates(Integer.parseInt(value));
                            break;
                        case "derrotas":
                            clube.setDerrotas(Integer.parseInt(value));
                            break;
                        case "gols_pro":
                            clube.setGolsPro(Integer.parseInt(value));
                            break;
                        case "gols_contra":
                            clube.setGolsContra(Integer.parseInt(value));
                            break;
                    }
                }
            }
        }
        
        return clube;
    }

    @Override
    public String toString() {
        return String.format("%s (%s) - %d pts, %d jogos", 
                           nome, sigla, getPontos(), getTotalJogos());
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Clube clube = (Clube) obj;
        return sigla != null ? sigla.equals(clube.sigla) : clube.sigla == null;
    }

    @Override
    public int hashCode() {
        return sigla != null ? sigla.hashCode() : 0;
    }
}

File: trabalho_3/cliente_javascript/js/models/Clube.js
/**
 * Classe que representa um clube de futebol
 */
class Clube {
    constructor(data = {}) {
        this.nome = data.nome || '';
        this.sigla = data.sigla || '';
        this.vitorias = data.vitorias || 0;
        this.empates = data.empates || 0;
        this.derrotas = data.derrotas || 0;
        this.gols_pro = data.gols_pro || 0;
        this.gols_contra = data.gols_contra || 0;
    }

    /**
     * Cria uma instância de Clube a partir de dados JSON do servidor
     * @param {Object} jsonData - Dados JSON do servidor
     * @returns {Clube} Nova instância de Clube
     */
    static fromJSON(jsonData) {
        return new Clube(jsonData);
    }

    /**
     * Converte a instância para um objeto JSON para envio ao servidor
     * @returns {Object} Objeto JSON
     */
    toJSON() {
        return {
            nome: this.nome,
            sigla: this.sigla,
            vitorias: this.vitorias,
            empates: this.empates,
            derrotas: this.derrotas,
            gols_pro: this.gols_pro,
            gols_contra: this.gols_contra
        };
    }

    /**
     * Calcula o total de jogos
     * @returns {number} Total de jogos
     */
    getTotalJogos() {
        return this.vitorias + this.empates + this.derrotas;
    }

    /**
     * Calcula os pontos (3 por vitória, 1 por empate)
     * @returns {number} Total de pontos
     */
    getPontos() {
        return (this.vitorias * 3) + this.empates;
    }

    /**
     * Calcula o saldo de gols
     * @returns {number} Saldo de gols
     */
    getSaldoGols() {
        return this.gols_pro - this.gols_contra;
    }

    /**
     * Retorna uma representação em string do clube
     * @returns {string} String formatada com informações do clube
     */
    toString() {
        return `${this.nome} (${this.sigla}) - ${this.getPontos()} pts, ${this.getTotalJogos()} jogos`;
    }
}
